### PR TITLE
Issue #8970: Check existence of ISO before opening properties window

### PR DESCRIPTION
**Pull request by Mikevin**
Existence of ISO file is checked before opening the properties windows to fix Dolphin crashing. GameList is refreshed to indicate missing file.

[Review on Reviewable](https://reviewable.io/reviews/dolphin-emu/dolphin/3073)

### CODE REVIEW

File: Source/Core/DolphinWX/GameListCtrl.cpp [hunk 1]
```diff
@@ -1011,8 +1011,15 @@ void CGameListCtrl::OnDeleteISO(wxCommandEvent& WXUNUSED (event))
 void CGameListCtrl::OnProperties(wxCommandEvent& WXUNUSED (event))
 {
 	const GameListItem* iso = GetSelectedISO();
-	if (!iso)
+	DiscIO::IVolume* OpenISO = DiscIO::CreateVolumeFromFilename(iso->GetFileName());
+
```
Comment (phire) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: Lint is complaining about the extra tab on this empty line.

File: Source/Core/DolphinWX/GameListCtrl.cpp [hunk 2]
```diff
@@ -1011,8 +1011,15 @@ void CGameListCtrl::OnDeleteISO(wxCommandEvent& WXUNUSED (event))
 void CGameListCtrl::OnProperties(wxCommandEvent& WXUNUSED (event))
 {
 	const GameListItem* iso = GetSelectedISO();
-	if (!iso)
+	DiscIO::IVolume* OpenISO = DiscIO::CreateVolumeFromFilename(iso->GetFileName());
```
Comment (phire) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: Could you use a unique pointer here, so it gets automatically deleted at the end of the function.

Comment (degasus) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: The !iso case will crash here.

**Comment by JosJuice** [comment]
I don't like that a volume object is being created just to be discarded right away. Would it be possible to check the volume object that's being created in the ISOProperties constructor instead?

File: Source/Core/DolphinWX/GameListCtrl.cpp [hunk 3]
```diff
@@ -1011,8 +1011,15 @@ void CGameListCtrl::OnDeleteISO(wxCommandEvent& WXUNUSED (event))
...
+	
+	if (!iso || !OpenISO)
+	{
+		delete OpenISO;
+		Update();
 		return;
+	}
+	delete OpenISO;
 
 	CISOProperties* ISOProperties = new CISOProperties(*iso, this);
```
Comment (phire) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: Not your code, but are we leaking the CISOProperties object here?

**Comment by phire** [comment]
Is silently updating the gamelist the best user experience here, if they are not looking closely they might miss the game disappearing and wonder why nothing happens and then why they can't find the game to open the properties again.

Perhaps an alert() dialogue?

**Comment by Mikevin** [comment]
Sorry guys, first PR to this (beautiful) repo.

I didn't want to create the volume object just to discard it right away but doing this in the ISOProperties constructor seems wrong as this window feels like it should be able to count on the existance of the file. Also not sure if it's nice to open a window and close it right away.

Would it be better to implement a way to check the existance from the GameListItem? Maybe in IsValid()?

**Comment by phire** [comment]
Another option would be to pass the volume object you create into the ISOProperties constructor so it doesn't have to create it a second time.

**Comment by JosJuice** [comment]
I'm not sure if the window even becomes visible if it's closed in the beginning of the constructor. Give it a try?

No, GameListItem isn't more appropriate. IsValid() doesn't get called at this point, and adding calls to it and creating a volume in it would be inappropriate for the same reasons as now.

File: Source/Core/DolphinWX/GameListCtrl.cpp (hunk 3 above)
Comment (phire) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: Nope. it looks like Wx takes ownership of the pointer and deletes it at some point after the window is closed.

An interesting design decision.

**Comment by Mikevin** [comment]
An alert dialogue and passing in the object sounds perfect. I'll try that, without the trailing ws this time.

File: Source/Core/DolphinWX/GameListCtrl.cpp [hunk 4]
```diff
@@ -1011,8 +1011,15 @@ void CGameListCtrl::OnDeleteISO(wxCommandEvent& WXUNUSED (event))
 void CGameListCtrl::OnProperties(wxCommandEvent& WXUNUSED (event))
 {
 	const GameListItem* iso = GetSelectedISO();
-	if (!iso)
+
+	if (!iso || !File::Exists(iso->GetFileName()))
+	{
+		WxUtils::ShowErrorDialog(
+			wxString::Format(_("Could not open %s. File might be renamed or deleted."),
```
Comment (JosJuice) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: There should be a "the" in front of "file".

**Comment by JosJuice** [comment]
The current version of the PR will catch cases of the file being moved, but not cases of the file being replaced with a file that exists but that isn't a valid volume. That's probably not going to happen often, though...

File: Source/Core/DolphinWX/GameListCtrl.cpp (hunk 4 above)
Comment (void-ghost) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: I think the correct form is "The file might have been renamed or deleted."

**Comment by lioncash** [comment]
The ideal way to accomplish this is to use a wxFileSystemWatcher to check for file existence in the background and remove entries from the list automatically if their corresponding file doesn't exist.

**Pull request closed without merging**